Use `ec_group_do_inverse_ord()` in SM2

romen opened the pull request:
This PR build on top of #6116 : 
- d6db841580 moves the `ec_group_do_inverse_ord()` declaration from `crypto/ec/ec_lcl.h` to a newly created `crypto/include/internal/ec_int.h`. The new header is meant to include *private* but *inter-submodule* accessible functions: up until now the SM2 submodule was using only public functions from the EC submodule.
- 313e72e39a updates the SM2 implementation to use `ec_group_do_inverse_ord()` instead of `BN_mod_inverse()`, mitigating a SCA leakage.

*It's 1.1.1 only.*

#### TODOs:
- [x] reshape commit `AuthorDate`s so github history is coherent with git commit history
- [x] gather feedback on the new commits
- [x] add documentation for `ec_group_do_inverse_ord()` in `crypto/include/internal/ec_int.h`
- [x] decide on documentation text: https://github.com/openssl/openssl/pull/6521#discussion_r197213329
- [x] rebase once #6116 is merged
- [x] rebase on master to fix some [-Wstrict-prototype] warnings
- [x] remove "WIP:" prefix for proper review

romen commented:
Ping to @mattcaswell and @bbbrumley

romen commented on crypto/sm2/sm2_sign.c:
```diff
@@ -11,6 +11,7 @@
 
 #include "internal/sm2.h"
 #include "internal/sm2err.h"
+#include "../ec/ec_lcl.h" /* ec_group_do_inverse_ord() */
```
@mattcaswell is this the right way to include stuff from the private API of the EC submodule?

mattcaswell replied:
No...if there are functions shared between internal sub-modules then they should be put in a header file in crypto/include/internal. You might have to create a new one if it doesn't already exist.

romen replied:
@mattcaswell So probably this latter change needs to be applied to #6116 before merging?

romen replied:
I mean, including a definition of `ec_group_do_inverse_ord()` in a header in `crypto/include/internal`

mattcaswell replied:
Well, you could do it either way. In some ways it makes more sense to do it here, i.e. don't make it into an internal include file until you actually need to use it. But whatever you think is easiest.

dot-asm replied:
Right, concern here is that relative path in #include can vary between implementations. To counter specifically this concern one can add `INCLUDE[sm2_sign.o]=../ec` to crypto/sm2/build.info and include ec_lcl.h.

mattcaswell replied:
> Right, concern here is that relative path in #include can vary between implementations.

That is *a* concern, but it is not my main concern. As a general principle we should not expose all of the internals of one sub-module to other sub-modules. There should be a clearly defined "API" between one sub-module and another. The internal include files define what it is reasonable for other sub-modules to have access to, and what it is not.

For that reason I do not think we should take that approach.

romen replied:
Ok, so let's get #6116 merged, given that it's ready, and then I'll move the prototype declaration where is more appropriate in this PR!

Thanks for the feedback!

romen commented:
I just rebased on the latest #6116 developments and added a new commit (2a6bdb3) which moves the `ec_group_do_inverse_ord()` declaration from `crypto/ec/ec_lcl.h` to a newly created `crypto/include/internal/ec_int.h`. The new header is meant to include private but inter-submodule accessible functions.

I then rebased the old commit that was including `crypto/ec/ec_lcl.h` breaking submodule separation, to instead include the new inter-submodule header.

As a side-note: *In my mind these 2 commits shouldn't be squashed together before merging.*

@mattcaswell @dot-asm would you care to provide more feedback?
- does the new header look good to you?
- does it need extra (`#ifndef`/`#define`) guards?
- are there conventions I am unaware of for naming these guards? 
- does this fix your previous concern(s)?
- is the build system tracking header files too? should the new internal header be listed somewhere?

romen commented:
*Added documentation for `ec_group_do_inverse_ord()` in `crypto/include/internal/ec_int.h`.*

romen commented on crypto/include/internal/ec_int.h:
```diff
@@ -0,0 +1,44 @@
+ *   of the output is not guaranteed.
+ * - When x is outside the range [1, EC_GROUP::order), the correctness
+ *   of the output is not guaranteed.
+ */
```
@bbbrumley Should I just go ahead and simply say:
```
 * This function expects that the EC_GROUP order is prime and that x is
 * included in the range [1, EC_GROUP::order).
 * In any other case, the correctness of its output is not guaranteed.
```

romen replied:
This should be the contract between the EC module and other submodules about this function, given that other submodules should probably not even know (or threat as an opaque object) what an `EC_METHOD` is and if it has a specialized implementation with looser requirements or not.

bbbrumley replied:
I would say `If the EC_GROUP order is even` so there is no ambiguity of which "group" we're referring to. `group->order` is the cardinality of the subgroup generated by G.

romen commented:
Rebasing this now that #6116 is merged.

romen commented on crypto/include/internal/ec_int.h:
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
```
@mattcaswell Is the copyright line ok? Or should it be
~~~
 * Copyright 2018-2018 The OpenSSL Project Authors. All Rights Reserved.
~~~

richsalz replied:
Just one year is fine.  Our script updates the copyrights on all files touched anyway so don't worry too much.

romen replied:
Thanks @richsalz

dot-asm commented on crypto/include/internal/ec_int.h:
```diff
@@ -0,0 +1,44 @@
+
+#  ifdef __cplusplus
+extern "C" {
+#  endif
```
This is internal header and it's never exposed to C++ compiler, hence there is no need for this. But if one chooses to have it, wouldn't after #include <openssl/ec.h> be the right spot?

romen replied:
> This is internal header and it's never exposed to C++ compiler, hence there is no need for this.

That's a very good point, and one I wondered about myself: I went for inclusion when I copied the basic structure form the recent `crypto/include/internal/sm2.h`.

My final rationale in keeping also what didn't make much sense to me was that this header was very recently added after a long review, and that reviewer definitely knew better!

mattcaswell replied:
> My final rationale in keeping also what didn't make much sense to me was that this header was very recently added after a long review, and that the reviewers definitely knew better!

No, the reviewers just missed it! :-(

romen replied:
@dot-asm I put #6554 to fix also the SM2 internal header

romen replied:
@mattcaswell It's not like reviewers are reviewing tons of code these days!! 😛 

BTW, I'm trying out the `git commit --fixup` trick you suggested: thanks, that's very useful!!

dot-asm replied:
That specific header was not added, but moved from public. Well, it's not really an excuse, just an explanation...

mattcaswell commented on crypto/include/internal/ec_int.h:
```diff
@@ -0,0 +1,36 @@
+/* Internal EC functions for other submodules: not for application use */
+
+#ifndef HEADER_OSSL_EC_INTERNAL_H
+# define HEADER_OSSL_EC_INTERNAL_H
```
Side note: We don't seem to have a consistent naming scheme for these guards. We should look at that - but this looks ok for now.

mattcaswell commented on crypto/include/internal/ec_int.h:
```diff
@@ -0,0 +1,36 @@
+ * - When the EC group order is otherwise composite, the correctness
+ *   of the output is not guaranteed.
+ * - When x is outside the range [1, EC_GROUP::order), the correctness
+ *   of the output is not guaranteed.
```
We should probably state what the return values are

romen replied:
I'll address return values as soon as I get feedback from @bbbrumley on https://github.com/openssl/openssl/pull/6521#discussion_r197213329.

He has a better understanding than me about the non-default paths when `EC_METHOD`s provide their own specific implementation of this functionality.

What I wrote here (or the alternative text) is definitely correct for the new default path, but given that the specialized versions have been written without documentation I don't want to make wrong assumptions: he looked at that code more than I did!

romen replied:
Latest revision explicitly states the return values and the caveats on reported success.

mattcaswell commented on crypto/include/internal/ec_int.h:
```diff
@@ -0,0 +1,36 @@
+ * - When x is outside the range [1, EC_GROUP::order), the correctness
+ *   of the output is not guaranteed.
+ */
+int ec_group_do_inverse_ord(const EC_GROUP *group, BIGNUM *res,
```
Would it be worth adding `__owur` here? (warn on unused result)

romen replied:
I'd say that annotation is a very good idea, always!

I'll look up how you are adding it on other functions to try and be consistent with the code style!

romen replied:
Added in the last revision.

romen commented:
About lifting the WIP, I'd like to finalize the documentation on the header before that::we might possibly need to discuss if the "contract" I wrote is too restrictive considering existing implementations and cornercase applications, and decide how to address that.

romen commented:
@mattcaswell @dot-asm @richsalz I have lifted the WIP, it should be now ready for proper review.

@bbbrumley could you once again comment on the current text for the documentation?

dot-asm commented:
Merged. Thanks!
